Show paste button in qr scanner dialog

nicodh opened the pull request:
resolves #3936

move "paste from clipboard" from settings context menu to own button

![image](https://github.com/user-attachments/assets/9330c0d1-822d-4351-8d69-ac02e2e7765e)

Another option would be: name the paste button "Import" and show the context menu onClick with the 2 options: 'load QR Code image' or 'paste from clipboard'
But buttons that show a context menu on click are a bit unusual

nicodh commented:
TODO: We have to adapt the e2e test for pasting the qr code! #4285

WofWca commented on packages/frontend/src/components/dialogs/QrCode.tsx:
```diff
@@ -235,6 +239,63 @@ export function QrCodeScanQrInner({
+  // callbacks when this React component has already been unmounted.
+  const handleScanResult = useCallback(
+    result => {
+      let unmounted = false
+
+      if (unmounted) {
+        return
+      }
+
+      handleScan(result.data)
+
+      return () => {
+        unmounted = true
+      }
```
So this is copy-pasted. Because `unmounted` doesn't appear to do anything. Just saying.

nicodh replied:
That was code that already existed. I removed it since I could not create a case, where the callback was called in an unmounted component, even with clicking "through" the overlay to close the dialog immediately
@Simon-Laux do you remember why that was added? 
The comment said:
>  // Additionally we have checks in place to make sure we're not firing any
 > // callbacks when this React component has already been unmounted.

nicodh replied:
If you consider this change to "risky" we could either readd it or postpone this PR for the current release

WofWca replied:
To me this looks like dead code. `unmounted` is never read after initialization, it's a new variable on each function execution, and it's only scoped to the function.

WofWca reviewed (commented):
This removes the "paste from clipboard" menu item from `QrReader`, but `QrReader` is [also used in `ReceiveBackupDialog`](https://github.com/deltachat/deltachat-desktop/blob/a5c02433976909fe01a64f3049c96d0d101ae9d8/packages/frontend/src/components/dialogs/SetupMultiDevice/ReceiveBackupDialog.tsx#L71). Does it mean it's now impossible to paste the backup QR?

r10s commented:
> Another option would be: name the paste button "Import" and show the context menu onClick with the 2 options: 'load QR Code image' or 'paste from clipboard'
But buttons that show a context menu on click are a bit unusual

let's stay with the paste button.

having that hidden at "import" again is not  obvious, if is also about the balance copy/paste. there could be a three dot menu beside the paste button for the other options, but that is another effort, and maybe not needed.

nicodh commented:
> This removes the "paste from clipboard" menu item from `QrReader`, but `QrReader` is [also used in `ReceiveBackupDialog`](https://github.com/deltachat/deltachat-desktop/blob/a5c02433976909fe01a64f3049c96d0d101ae9d8/packages/frontend/src/components/dialogs/SetupMultiDevice/ReceiveBackupDialog.tsx#L71). Does it mean it's now impossible to paste the backup QR?

True, that's another use case not covered in the Issue since the left footer button is already set there. So maybe we should move the Troubleshoot link to another place? @r10s

WofWca commented:
A simple hack would be to keep the "paste from clipboard" menu item just for the Backup dialog. I.e. to add a prop `hidePasteFromClipboardMenuOption` to `QrReader`.

Simon-Laux commented:
I'd say keep the paste button in the menu or move trouble shooting up to the description.
BTW: I think we can remove the experimental label and version requirement

r10s commented:
i'd keep the menu as is, that seems to be the easiest approach.

for the "experimental" hint: we're staying with that for so long, let's do not remove that last minute, it is fine.  idk, what the exact state of iroh is, let's see that for 1.50 and maybe get it out on all platforms

Simon-Laux commented:
> also for reasons, regarding backup/backup2 hickup

wouldn't we need to upgrade the version then or is the first iroh/backup(without the 2) still supported?

r10s commented:
> wouldn't we need to upgrade the version then or is the first iroh/backup(without the 2) still supported?

i think it is fine, as long as desktop checks for DC_QR_BACKUP as well as DC_QR_BACKUP2 at all places, on android we forgot one place, and that resulted in a degradation, see https://github.com/deltachat/deltachat-android/pull/3370

nicodh commented:
I readded the "paste from clipboard" item to settings menu

I also refactored the implementation to avoid duplicate code by using [useImperativeHandle](https://react.dev/reference/react/useImperativeHandle)
I removed some existing code since - see comment above

nicodh commented:
> TODO: We have to adapt the e2e test for pasting the qr code! #4285

Done

WofWca commented on packages/frontend/src/components/dialogs/QrCode.tsx:
```diff
@@ -228,8 +233,8 @@ export function QrCodeScanQrInner({
           openAlertDialog({
             message: error.message || error.toString(),
           })
+          processingQrCode.current = false
```
Why has this been moved? The logic doesn't look right now.
And a copy-pasted version of this code is also in `ReceiveBackupDialog.tsx`, where it remains unchanged.

nicodh replied:
if processQr is successfull it calls onDone which sets processingQrCode.current to false, but itf it fails it has to be set in the catch clause.

So both is correct, but I agree it is confusing to have it one time like this and one time like that

nicodh replied:
Moved it back

WofWca commented on packages/frontend/src/components/QrReader/index.tsx:
```diff
@@ -129,25 +158,40 @@ export default function QrReader({ onError, onScan }: Props) {
+      if (typeof result === 'string') {
+        onScan(result)
+      } else {
+        try {
+          onScan(result.data)
+        } catch (error) {
+          console.error('Error while processing clipboard content:', error)
+        }
```
If an error occurs here, `userFeedback` doesn't get called. I believe this is not the case in the pre-MR version, where if `handleScanResult` throws, we do call `userFeedback`. Is this what we want?

nicodh replied:
onScan has its own error handling anyway so this is not needed

nicodh replied:
I pass the error now always to onError so the parent component shows an AlertDialog

WofWca commented on packages/frontend/src/components/QrReader/index.tsx:
```diff
@@ -92,7 +94,34 @@ async function base64ToImageData(base64: string): Promise<ImageData> {
+    const imageData = await base64ToImageData(base64)
+    const result = scanQrCode(imageData.data, imageData.width, imageData.height)
+    if (result) {
+      return result
```
Nitpick: this function is only used in one place, and only the `data` property of this return value is used. I think it'd be simpler to make it return `result.data` only, so there is no need to check `typeof` of the return value if this function.

nicodh replied:
That's true - it is a leftover of the previous handleScanResult which I don't know of why it was neccessary (see my question to simon). There the result was stripped down to result.data

I will reduce that...

WofWca commented on packages/frontend/src/components/QrReader/index.tsx:
```diff
@@ -129,25 +158,40 @@ export default function QrReader({ onError, onScan }: Props) {
+      if (typeof result === 'string') {
+        onScan(result)
+      } else {
+        try {
+          onScan(result.data)
+        } catch (error) {
+          console.error('Error while processing clipboard content:', error)
+        }
       }
-
-      onScan(result.data)
-
-      return () => {
-        unmounted = true
+    } catch (error) {
+      userFeedback({
+        type: 'error',
+        text: `${tx('qrscan_failed')}: ${error}`,
+      })
+    }
+  }, [onScan, tx, userFeedback])
+
+  /**
+   * used to expose the handlePasteFromClipboard
+   * function to the parent component
+   */
+  useImperativeHandle(
+    ref,
+    () => {
+      return {
+        pasteFromClipboard() {
+          handlePasteFromClipboard()
+        },
       }
     },
-    [onScan]
+    [handlePasteFromClipboard]
   )
```
Nitpick: given how little `handlePasteFromClipboard` does (especially if you take my other commets into account), I personally wouldn't make it an imperative handle. Instead I'd just completely separate it from the `QrReader` component.

WofWca reviewed (changes requested):
With my non-nitpick comments addressed or dismissed, it's good to merge.

I tested it a bit, and looked at the code. It seems fine otherwise.
Though NGL, it's a bit hard to review, with refactoring changes and functional changes being mixed.

nicodh commented:
@r10s since I refactor this now anyway: when do we show the red alert (in our code is named "userFeedback") and when a normal warning dialog?
Since when a dialog is opened, the red message is behind the backdrop which is not so nice:

![image](https://github.com/user-attachments/assets/4cbbb758-c6a4-4c5a-a165-0447a6236649)

the "AlertDialog" would look like that:

![image](https://github.com/user-attachments/assets/a8982e35-a4c1-412e-b5c2-bbf4e9ad6ed8)

which is not really what I would expect from a "AlertDialog" (butb that is probably another issue)

nicodh commented:
When scanning a qr code for backup import there are even 2 backdrops over the user feedback, so it is almost invisible

![image](https://github.com/user-attachments/assets/f5e8f2ed-dc25-46b5-949f-8fc832a090f0)

WofWca commented:
FYI that `user-feedback` toast is ancient. Last time it was changed was in 2019: b0134fef743e077a5af3f63174fa836d8ac12c24.

WofWca reviewed (approved):
So, there are a lot of changes in this MR, both functional and refactoring. Not all are strictly related to the title of the MR,

I have verified that the moved `fileToBase64` and `base64ToImageData` functions are unchanged. `qrCodeFromClipboard` was moved from `handlePasteFromClipboard` and it's also largely unchanged. The only thing different is that it errors out when the scan result's string is empty, which probably makes sense (but I'm not sure if this can happen).

About the removal of `setError(true)` from `handleError`, and the rename of it to `setCameraAccessError`. I haven't checked what it does exactly.

But what I noticed is that we now handle all scanning errors in one place, in the parent component, which is probably good. But we also invoke the callback after an error is thrown in `worker.postMessage(imageData)`, and `postMessage` gets called every 250ms. But I guess this can't happen often.

Another things is as was said above, now an `AlertDialog` is used instead of the `userFeedback` toast, which is better (at least in common cases).

As to the rest of the changes (that is the "add 'paste' button"): they look good, apart from that one suggestion I left!

nicodh commented:
> setCameraAccessError

this error is set if startStream fails so I named it a bit more meaningful

I think we won't merge this into the current release (which will be published tomorrow hopefully)

nicodh commented on packages/frontend/src/components/QrReader/index.tsx:
```diff
@@ -158,47 +82,18 @@ export default function QrReader({ onError, onScan }: Props) {
         onError(error.toString())
       }
-
-      setError(true)
```
Yes but that resulted in showing the camera_access_failed error in the UI which I consider wrong since it's a generic errorHandler (for example after an error in handleFileInputChange)
